sw/device/silicon_creator/lib/drivers/flash_ctrl.c

```diff
@@ -142,48 +176,66 @@ void flash_ctrl_status_get(flash_ctrl_status_t *status) {
 }
 
 rom_error_t flash_ctrl_read(uint32_t addr, uint32_t word_count,
-                            flash_ctrl_partition_t region, uint32_t *data) {
-  // Start the read transaction, the value of `erase_type` doesn't matter.
-  RETURN_IF_ERROR(transaction_start(addr, word_count, region,
-                                    kFlashCtrlErasePage,
-                                    FLASH_CTRL_CONTROL_OP_VALUE_READ));
+                            flash_ctrl_partition_t partition, uint32_t *data) {
+  RETURN_IF_ERROR(transaction_start((transaction_params_t){
+      .addr = addr,
+      .op_type = FLASH_CTRL_CONTROL_OP_VALUE_READ,
+      .partition = partition,
+      .word_count = word_count,
+      // Does not apply to read transactions.
+      .erase_type = kFlashCtrlEraseTypePage,
+  }));
   fifo_read(word_count, data);
   return wait_for_done();
 }
 
 rom_error_t flash_ctrl_prog(uint32_t addr, uint32_t word_count,
-                            flash_ctrl_partition_t region,
+                            flash_ctrl_partition_t partition,
                             const uint32_t *data) {
-  uint32_t window_offset = addr % FLASH_CTRL_PARAM_REG_BUS_PGM_RES_BYTES;
-
+  enum {
+    kWindowWordCount =
+        FLASH_CTRL_PARAM_REG_BUS_PGM_RES_BYTES / sizeof(uint32_t),
+  };
+
+  // Find the number of words that can be written in the first window.
+  uint32_t window_word_count =
+      kWindowWordCount - ((addr / sizeof(uint32_t)) % kWindowWordCount);
   while (word_count > 0) {
     // Program operations can't cross window boundaries.
-    uint32_t max_bytes = FLASH_CTRL_PARAM_REG_BUS_PGM_RES_BYTES - window_offset;
-    uint32_t write_size = max_bytes / sizeof(uint32_t);
-    write_size = word_count < write_size ? word_count : write_size;
-    window_offset = 0;
-
-    // Start the program transaction, the value of `erase_type` doesn't matter.
-    RETURN_IF_ERROR(transaction_start(addr, write_size, region,
-                                      kFlashCtrlErasePage,
-                                      FLASH_CTRL_CONTROL_OP_VALUE_PROG));
-
-    fifo_push(write_size, data);
+    window_word_count =
+        word_count < window_word_count ? word_count : window_word_count;
+
+    RETURN_IF_ERROR(transaction_start((transaction_params_t){
+        .addr = addr,
+        .op_type = FLASH_CTRL_CONTROL_OP_VALUE_PROG,
+        .partition = partition,
+        .word_count = window_word_count,
+        // Does not apply to program transactions.
+        .erase_type = false,
```

Comment:
Is the false assignment here correct? I think this should be an enum value.

Reply:
Good catch! I think initially I had this as a bool. Updated `flash_ctrl_read()` but missed this one. Thanks!